core/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java

```diff
@@ -354,15 +352,25 @@ private void addMappers(Collection<ObjectMapper> objectMappers, Collection<Field
         mapperService.addMappers(objectMappers, fieldMappers);
     }
 
-    public MergeResult merge(Mapping mapping, boolean simulate, boolean updateAllTypes) {
+    public void merge(Mapping mapping, boolean updateAllTypes) {
         try (ReleasableLock lock = mappingWriteLock.acquire()) {
-            final MergeResult mergeResult = new MergeResult(simulate, updateAllTypes);
+            // first simulate to only check for conflicts
+            MergeResult mergeResult = new MergeResult(true, updateAllTypes);
             this.mapping.merge(mapping, mergeResult);
-            if (simulate == false) {
-                addMappers(mergeResult.getNewObjectMappers(), mergeResult.getNewFieldMappers(), updateAllTypes);
-                refreshSource();
+            if (mergeResult.hasConflicts()) {
+                throw new MergeMappingException(mergeResult.buildConflicts());
             }
-            return mergeResult;
+
+            // then check conflicts with other types
+            // if there is a single type then we are implicitly updating all types
+            mapperService.checkMappersCompatibility(mapping, updateAllTypes || mapperService.types().size() == 1);
+
+            // finally apply the mapping update for real
+            mergeResult = new MergeResult(false, updateAllTypes);
```

Comment:
why do we need to merge this again since we are still holding on to the lock? I don't necessarily understand why this is helping us as well but that might just be because I don't know this code very well.